src/USER-PHONON/dynamical_matrix.cpp

```diff
@@ -0,0 +1,491 @@
+//
+// Created by charlie sievers on 6/21/18.
+//
+
+#include <mpi.h>
+#include <cstdlib>
+#include "dynamical_matrix.h"
+#include "atom.h"
+#include "modify.h"
+#include "domain.h"
+#include "comm.h"
+#include "group.h"
+#include "force.h"
+#include "math_extra.h"
+#include "memory.h"
+#include "bond.h"
+#include "angle.h"
+#include "dihedral.h"
+#include "improper.h"
+#include "kspace.h"
+#include "update.h"
+#include "neighbor.h"
+#include "pair.h"
+#include "timer.h"
+#include "finish.h"
+
+
+using namespace LAMMPS_NS;
+enum{REGULAR,ESKM};
+
+/* ---------------------------------------------------------------------- */
+
+DynamicalMatrix::DynamicalMatrix(LAMMPS *lmp) : Pointers(lmp), fp(NULL)
+{
+    external_force_clear = 1;
+}
+
+/* ---------------------------------------------------------------------- */
+
+DynamicalMatrix::~DynamicalMatrix()
+{
+    if (fp && me == 0) fclose(fp);
+    memory->destroy(dynmat);
+    memory->destroy(final_dynmat);
+    fp = NULL;
+}
+
+/* ----------------------------------------------------------------------
+   setup without output or one-time post-init setup
+   flag = 0 = just force calculation
+   flag = 1 = reneighbor and force calculation
+------------------------------------------------------------------------- */
+
+void DynamicalMatrix::setup()
+{
+    // setup domain, communication and neighboring
+    // acquire ghosts
+    // build neighbor lists
+    if (triclinic) domain->x2lamda(atom->nlocal);
+    domain->pbc();
+    domain->reset_box();
+    comm->setup();
+    if (neighbor->style) neighbor->setup_bins();
+    comm->exchange();
+    comm->borders();
+    if (triclinic) domain->lamda2x(atom->nlocal+atom->nghost);
+    domain->image_check();
+    domain->box_too_small_check();
+    neighbor->build(1);
+    neighbor->ncalls = 0;
+    neighbor->every = 2;                       // build every this many steps
+    neighbor->delay = 1;
+    neighbor->ago = 0;
+    neighbor->ndanger = 0;
+
```

Comment:
Here is an idea for how you can add group support:
- count the number of atoms in the group and store it as a class member. that number replaces `atom->natoms` to determine the per-atom loop count in the computation (nelem)
- allocate a vector `idxmap[nelem]` and a vector `localmap[atom->nmax]` on every MPI rank and zero them
- on each MPI rank do a loop from 0 to `atoms->nlocal`(exclusive) and if the loop index i is member of the group add `atom->tag[i]` to localmap[]
- on MPI rank zero, copy the non-zero elements of localmap[] to idxmap[]
- now loop over all MPI ranks and have them send their localmap to the one on rank 0, and append the non-zero elements to idxmap[]
- now sort idxmap[] (if desired) and broadcast it to all MPI ranks. use the content of idxmap instead of the loop from 1 to `atom->natoms`(inclusive) for the loops in the dynamical matrix calculation. 
- when outputting the dynamical matrix, also provide this list, so that the matrix elements can be identified by atom id and direction.

Reply:
I believe I already did this exact solution.

Reply:
nope. my suggestion will reduce the amount of memory used, yours will always require storage of the dimension natoms.